tools/wptrunner/wptrunner/executors/executorcontentshell.py

```diff
@@ -0,0 +1,211 @@
+# mypy: allow-untyped-defs
+
+from .base import RefTestExecutor, RefTestImplementation, CrashtestExecutor, TestharnessExecutor
+from .protocol import ConnectionlessProtocol
+from time import time
+from queue import Empty
+from base64 import b64encode
+from os import linesep
+import json
+
+class CrashError(BaseException):
+    pass
+
+class ContentShellProtocol(ConnectionlessProtocol):
+    """This class represents the protocol used by content_shell in protocol mode.
+    For more details, see:
+    https://chromium.googlesource.com/chromium/src.git/+/HEAD/content/web_test/browser/test_info_extractor.h
+    """
+    # Marker sent by content_shell after blocks.
+    eof_marker = "#EOF" + linesep
+
+    def teardown(self):
+        # Close the queue properly to avoid broken pipe spam in the log.
+        self.browser.stdin_queue.close()
+        self.browser.stdin_queue.join_thread()
+
+    def send_command(self, command):
+        """Sends a single `command`, i.e. a URL to open, to content_shell.
+        """
+        self.browser.stdin_queue.put((command + linesep).encode("utf-8"))
+
+    def read_errors(self):
+        """Reads the entire content of the stderr queue as is available right now (no blocking).
+        """
+        result = ""
+
+        while not self.browser.stderr_queue.empty():
+            # There is no potential for race conditions here because this is the only place
+            # where we read from the stderr queue.
+            error = self.browser.stderr_queue.get()
+            result += error.decode("utf-8", "replace") + "\n"
+
+        return result
+
+    def read_line(self, deadline=None, encoding=None, errors="strict", raise_crash=True):
+        """Reads a single line from the stdout queue. The read must succeed before `deadline` or
+        a TimeoutError is raised. The line is returned as a bytestring or optionally with the
+        specified `encoding`. If `raise_crash` is set, a CrashError is raised if the line
+        happens to be a crash message.
+        """
+        current_time = time()
+
+        if deadline and current_time > deadline:
+            raise TimeoutError()
+
+        try:
+            line = self.browser.stdout_queue.get(True,
+                    deadline - current_time if deadline else None)
+
+            if raise_crash and line.startswith(b"#CRASHED"):
+                raise CrashError()
+        except Empty:
+            raise TimeoutError()
+
+        return line.decode(encoding, errors) if encoding else line
+
+    def is_alive(self):
+        """Checks if content_shell is still working by running a test on `about:blank` with
+        a short 2 second timeout.
+        """
+        try:
```

Comment:
Hmm, this seems like it's also affecting state. Really `is_alive` should be safe to call at any time. In this case maybe it makes sense to check that the queues are still open, or similar.

Reply:
I've changed it so that it checks if the IO pipes are still open. The problem with this is that we cannot know if the process is still responsive. Unfortunately, I don't know of a way to check that for content_shell without affecting state.